Chapter? Bayesian Optimization (#557)
mlr-org/mlr3book: book/bayesian_optimization.qmd

  A prime example is given by hyperparameter optimization where we configure a learner with a hyperparameter configuration and estimate its generalization performance via resampling.
  Moreover, hyperparameter optimization can be extremely expensive and therefore calls for very sample efficient optimization algorithms.
  This chapter presents an introduction to Bayesian optimization (also called model-based optimization) within the mlr3 ecosystem.
  We start with the general black box optimization setting and proceed to introduce the building blocks of Bayesian optimization - a class of sample efficient iterative global optimization algorithms.
  After that we show that Bayesian optimization can be used out of the box for hyperparameter optimization of machine learning algorithms.
  We further show how multi-objective Bayesian optimization can be performed and briefly discuss noisy objective functions.
  The chapter closes with comments on robustness, safety nets, defaults and practical considerations regarding Bayesian optimization algorithms.
---

# Bayesian Optimization {#sec-bayesian-optimization}

```{r}

> set.seed(1005)

Reviewer: Update once chapter position fixed

```

{{< include _setup.qmd >}}

Black box optimization considers the optimization of a so-called black box function whose structure and analytical description is unknown, unexploitable or non-existent.
As a result, the only observable information is the output value of the function given an input value.
Usually, the objective function is subject so some box constraint, i.e., in the case of numeric variables the domain is assumed to be a hyperrectangle.

> Examples for black box optimization problems range from real-life experiments, e.g., crash tests or chemical reactions to computer simulations.

Reviewer: Some of this is already covered in C4, as you note below, so would instead linking back to C4 in the first line and then generalising this to black-box more generally

Hyperparameter optimization (HPO, see @sec-optimization) is a prime example for black box optimization.
Recall that in HPO, we configure a learner with a hyperparameter configuration and evaluate the learner via a resampling technique to measure its performance with the goal to find the optimal hyperparameter configuration.
In general no analytical description for this mapping from a hyperparameter configuration to performance exists and gradient information is not available.
Besides, evaluating the performance of a learner can take a substantial amount of time, making HPO an expensive black box optimization problem.

Many algorithm classes exists that can be used for black box optimization that differ in how they tackle this optimization problem.

> In general, most black box optimizers work iterative, i.e., they propose new points for evaluation by making use of the information collected during the evaluation of previous points.

Reviewer: ```suggestion
In general, most black box optimizers work iteratively, i.e., they propose new points for evaluation by making use of the information collected during the evaluation of previous points.
```

Reviewer: I think this might have been covered in C4

Evolutionary strategies for example maintain a so-called population and generate new points for evaluation by choosing parents from that population and performing recombination operators such as mutation and crossover to generate the offspring of the next generation.
In general, evolutionary strategies are very suited for black box optimization, however, if the overhead of the evaluation of the black box function becomes large (e.g., as in HPO), sample efficiency of an optimizer becomes highly relevant.
Bayesian optimization (BO) describes a class of sample efficient iterative global black box optimization algorithms that make use of a so-called surrogate trained on observed data modelling the black box function.
Based on this (non-linear) regression model and a so-called acquisition function quantifying the attractiveness of each point in the search space, the next candidate point is obtained that is to be evaluated next.
After evaluating this candidate, the surrogate model is updated or re-trained and the process repeats itself until a termination criteria is met.

> Often, using BO results in very good optimization performance, especially if the overhead of the black box evaluation becomes large and optimization budget is tight.

Reviewer: I think it needs to be clearer why we are including this chapter before going into some complex terminology. Personally this is an area I'm very interested in learning more about, but this intro makes it seem complicated and doesn't tell me in simple terms *why* I should care

> This chapter is of general relevance to users concerned with sample efficient black box optimization and HPO.

Reviewer: As above, this is a bit too complicated for the 'why'

In the following, we will give a brief general introduction to black box optimization making use of the `r bbotk`\index{bbotk} package.
We then show how BO can be performed within the mlr3 ecosystem making use of the `r mlr3mbo`\index{mlr3mbo} package.
Detailed introductions to black box optimization and BO are given in @hpo_practical, @hpo_automl and @garnett_2022.

## Black Box Optimization {#sec-black-box-optimization}

The `r bbotk` package is the workhorse package for black box optimization within the mlr3 ecosystem.
At the heart of the package are the R6 classes:

* `r ref("OptimInstanceSingleCrit")` and `r ref("OptimInstanceMultiCrit")`, which are used to construct an optimization 'instance' which describes the optimization problem and stores the results; and
* `r ref("Optimizer")` which is used to get and set optimization algorithms.

This should sound very familiar.
Indeed, in @sec-optimization, the classes `r ref("TuningInstanceSingleCrit")`, `r ref("TuningInstanceMultiCrit")` and `r ref("Tuner")` were already introduced.

> Those three R6 classes essentially build upon `r ref("OptimInstanceSingleCrit")`, `r ref("OptimInstanceMultiCrit")` and `r ref("Optimizer")` which are the more general black box optimization base classes.

Reviewer: Would suggest rephrasing to make simpler as: i) in C4 we saw classes Tuner...; ii) these are special cases of Optim... from bbotk.

Our running example will be to optimize the following sinusoidal function which is characterized by two local minima and one global minimum:

> $f: [0, 1] \rightarrow \mathbb{R}, x \mapsto 2x + \sin(14x)$.

Reviewer: Reference figure below here

> The local optima can pose a risk for many optimization algorithms to get stuck.

Reviewer: Reference fig below, point to local minima and briefly explain why this means algs might get stuck

> At the core of an `r ref("OptimInstanceSingleCrit")` lies an `r ref("Objective")` function wrapping the function, domain and codomain (see also #sec-defining-search-spaces).

Reviewer: Broken link to sec-defining...

In the following, we will use `r ref("ObjectiveRFunDt")`:

```{r}
sinus_1D = function(xdt) {
  y = 2 * xdt$x * sin(14 * xdt$x)
  data.table(y = y)
}
```

An `r ref("Objective")` always requires the specification of the domain and codomain in the form of a `r ref("ParamSet")`:

```{r}
domain = ps(x = p_dbl(lower = 0, upper = 1))
codomain = ps(y = p_dbl(tags = "minimize"))
objective = ObjectiveRFunDt$new(sinus_1D,
  domain = domain, codomain = codomain)
```

We can proceed to visualize the sinusoidal function by generating a grid of points on which we evaluate the function:

```{r}
xydt = generate_design_grid(domain, resolution = 1001)$data
xydt[, y := objective$eval_dt(xydt)$y]
```

```{r}
#| echo: true
#| label: fig-bayesian-optimization-sinusoidal
#| fig-cap: Visualization of the sinusoidal function.
#| fig-alt: FIXME
library(ggplot2)
ggplot(aes(x = x, y = y), data = xydt) +
  geom_line() +
  theme_minimal()
```

The global minimum is located at `x = 0.7918238`:

```{r}

> sinus_1D(data.table(x = 0.7918238))

Reviewer: What does this function do?

```

By wrapping the objective function in an `r ref("OptimInstanceSingleCrit")` we can proceed to optimize it, e.g., via a simple random search.

```{r, output=FALSE}
instance = OptimInstanceSingleCrit$new(objective,
  terminator = trm("evals", n_evals = 20))
optimizer = opt("random_search", batch_size = 20)
optimizer$optimize(instance)
```

```{r}
instance$archive$best()
```

To see all available optimizers that can be used for black box optimization, you can inspect the following dictionary:

```{r}
as.data.table(mlr_optimizers)
```

> Note that evolutionary strategies are available within the mlr3 ecosystem via the `miesmuschel` package which we will not cover in this chapter.

Reviewer: link to GH

## Building Blocks of Bayesian Optimization {#sec-bayesian-optimization-blocks}

> Bayesian optimization (BO) is an iterative optimization algorithm that makes use of a so-called surrogate to model the unknown black box function.

Reviewer: what's a surrogate?

After having observed an initial design of observations, the surrogate model is trained on all data points observed so far and a so-called acquisition function is used to determine which point of the search space is a promising candidate that should be evaluated next.
The acquisition function relies on the predictions of the surrogate model and requires no evaluation of the true black box function and therefore is comparably cheap to optimize.
The acquisition function should balance exploration and exploitation of the BO algorithm, i.e., we want to exploit knowledge about regions where we observed that performance is good and the surrogate model has low uncertainty but also want to make sure that we do not miss crucial regions and therefore also want to explore into regions where the uncertainty of the surrogate model is high.

> After having evaluated the next candidate, the process repeats itself until a given termination criteria is met.

Reviewer: Sorry but this is too complicated to follow. Either break it up with examples, use some maths, or clearly define terms upfront before discussing how they work

Most BO flavors therefore follow a simple loop:

    1. Fit the surrogate model on all observations made so far.
    2. Optimize the acquisition function to find the next candidate that should be evaluated.
    3. Evaluate the next candidate.

> Note that we speak of BO flavors, as BO is highly modular, i.e., users can choose different types of surrogate models, acquisition functions and acquisition function optimizers to their liking.

Reviewer: I don't fully know what the 'note' is

> `r mlr3mbo` makes BO available within the mlr3 ecosystem.

Reviewer: New line

At the heart of the package are the two R6 classes `r ref("OptimizerMbo")` and `r ref("TunerMbo")` which can be configured with respect to their `r ref("loop_function")` (determining the general loop structure of the BO algorithm), `r ref("Surrogate")` (surrogate model), `r ref("AcqFunction")` (acquisition function) and `r ref("AcqOptimizer")` (acquisition function optimizer).

### The Initial Design

Before we can fit a surrogate model, we need data.

> The initial design refers of the first set of points the black box function was evaluated on before BO algorithms start their loop.

Reviewer: ```suggestion
The initial design refers to the first set of points the black box function was evaluated on before BO algorithms start their loop.
```

  geom_hline(yintercept = qs, linetype = 2) +
  labs(title = "LHS Design", x = expression(x[1]), y = expression(x[2])) +
  theme_minimal() +
  xlim(c(0, 1)) +
  ylim(c(0, 1))
g_lhs = ggMarginal(g_lhs, type = "histogram", bins = 10)

g_sobol = ggplot(aes(x = x1, y = x2), data = sobol_design) +
  geom_point(size = 3L) +
  geom_vline(xintercept = qs, linetype = 2) +
  geom_hline(yintercept = qs, linetype = 2) +
  labs(title = "Sobol Design", x = expression(x[1]), y = expression(x[2])) +
  theme_minimal() +
  xlim(c(0, 1)) +
  ylim(c(0, 1))
g_sobol = ggMarginal(g_sobol, type = "histogram", bins = 10)

grid.arrange(g_random, g_grid, g_lhs, g_sobol, nrow = 2, ncol = 2)
```

We observe that a random design does not cover the domain well and simply by chance samples points close to each other.
A grid design will always result in points being marginally equidistant from their nearest neighbour and does not cover the domain well.
In contrast, an LHS design will result in good coverage of the domain and even for a small number of samples the marginal distributions will be perfectly uniform (i.e., in the plot, in each row and column there is exactly one sample).

> A Sobol design has a similar goal in mind but must not guarantee this for a small number of samples.

Reviewer: I don't fully understand this description

However, constructing a Sobol design can be done much more efficiently than an LHS design, especially if the number of samples and dimensions grows.
Moreover, a Sobol design has better coverage properties than an LHS design if the number of dimensions grows large.

Of course, one can always evaluate a custom initial design in the form of a `data.table`.
Coming back to our running example, we will now evaluate the following custom design:

```{r}
instance = OptimInstanceSingleCrit$new(objective,
  terminator = trm("evals", n_evals = 20))
design = data.table(x = c(0.1, 0.3, 0.65, 1))
instance$eval_batch(design)
```

> ### Loop Function

Reviewer: Can you give a concrete example in code, pseudo-code, or as a figure?

The `r ref("loop_function")` determines the behavior of the BO algorithm on a global level, i.e., how the subroutine should look like that is performed at each iteration.

To get an overview of readily available loop functions, the following dictionary can be inspected:

```{r}
as.data.table(mlr_loop_functions)
```

Technically, all `loop_function`s are members of the `S3` class `r ref("loop_function")` (and are actually simply `r ref("base::function", "functions")` with some attributes set).

For sequential single-objective black box optimization, the Efficient Global Optimization (EGO) algorithm [@jones_1998] can be considered the reference algorithm.
In `r mlr3mbo`, the EGO algorithm is implemented via the `r ref("mlr_loop_functions_ego", "bayesopt_ego")` loop function.
After having made some assertions and safety checks, and having evaluated the initial design, `bayesopt_ego` essentially repeatedly performs the following steps:

1. `acq_function$surrogate$update()`: update the surrogate model
2. `acq_function$update()`: update the acquisition function
3. `acq_optimizer$optimize()`: optimize the acquisition function to yield a new candidate

> ### Surrogate

Reviewer: Whole section a bit hard to follow, don't have a good suggestion right now for how best to improve but basically there's just a lot of new and technical content

A surrogate encapsulates a regression learner that models the unknown black box function based on observed data.
In `r mlr3mbo`, `r ref("SurrogateLearner")` and `r ref("SurrogateLearnerCollection")` are the higher-level R6 classes which should be used to construct a surrogate, inheriting from the base `r ref("Surrogate")` class.

As a learner, any `r ref("LearnerRegr", "regression learner")` from `r "mlr3"` can be used, however, most acquisition functions require both a mean and a variance prediction (therefore not all learners are suitable for all scenarios).
Typical choices include:

> * A `r ref("mlr3learners::mlr_learners_regr.km", "Gaussian Process")` for low dimensional numeric search spaces

Reviewer: ```suggestion
* A Gaussian process (`r ref("mlr_learners_regr.km")`) for low dimensional numeric search spaces
```

Reviewer: we're removing the "text" argument from `ref`

> * A `r ref("mlr3learners::mlr_learners_regr.ranger", "Random Forest")` for higher dimensional mixed (and / or hierarchical) search spaces

Reviewer: ```suggestion
* A random forest for higher dimensional mixed (and / or hierarchical) search spaces
```
We implement several so shouldn't direct users to just one

> We will not dive into details regarding Gaussian Processes here.

Reviewer: ```suggestion
```

A detailed introduction to Gaussian Processes can be found in @williams_2006 and in-depth focus to Gaussian Processes in the context of surrogate models in BO is given in @garnett_2022.

A `SurrogateLearner` can be constructed via:

```{r}

> surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS", control = list(trace = FALSE)))

Reviewer: ```suggestion
surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS",
  control = list(trace = FALSE)))
```

```
Here, we use a Gaussian Process with Matérn 5/2 kernel, optimizing the marginal likelihood via BFGS and set `trace = FALSE` to prevent too much output during fitting.

When using a `Surrogate` interactively, i.e., outside of an `OptimizerMbo` or `TunerMbo`, the `archive` of the instance must be specified:

```{r}

> surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS", control = list(trace = FALSE)), archive = instance$archive)

Reviewer: ```suggestion
surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS",
  control = list(trace = FALSE)), archive = instance$archive)
```

Note that when using a Gaussian Process that interpolates the training data, the variance prediction is zero for training data.
The initial design is given by the black points and the true sinusoidal function by the black line.

```{r}
xydt[, c("mean", "variance") := surrogate$predict(xydt[, surrogate$x_cols, with = FALSE])]
```

```{r}
#| echo: true
#| label: fig-bayesian-optimization-mean-variance
#| fig-cap: Mean and variance prediction of the Gaussian Process surrogate model trained on an initial design of four points (black). Ribbons represent the mean plus minus the variance prediction.
#| fig-alt: FIXME
ggplot() +
  geom_point(aes(x = x, y = y), size = 2, data = instance$archive$data) +
  geom_line(aes(x = x, y = y), data = xydt) +
  geom_line(aes(x = x, y = mean), colour = "steelblue", linetype = 2, data = xydt) +
  geom_ribbon(aes(x = x, min = mean - variance, max = mean + variance), fill = "steelblue", colour = NA, alpha = 0.1, data = xydt) +
  theme_minimal()
```

### Acquisition Function

Based on a surrogate, an acquisition function quantifies the attractiveness of each point of the search space if it were to be evaluated in the next iteration.

> A popular example is given by the Expected Improvement [@jones_1998]:

Reviewer: This is great

  terminator = trm("stagnation", iters = 100, threshold = 1e-5),
  acq_function = acq_function)
```

In this example we use the DIRECT algorithm provided by the `nloptr` package to optimize the Expected Improvement.
We will terminate the acquisition function optimization if we no longer improve at least `1e-5` over a history of `100` iterations.

```{r}
candidate = acq_optimizer$optimize()
candidate
```

@fig-bayesian-optimization-ei-argmax visualizes the location of the next candidate found by maximizing the Expected Improvement.

```{r}
#| echo: true
#| label: fig-bayesian-optimization-ei-argmax
#| fig-cap: Next candidate point (darkred) obtained by maximizing the Expected Improvement based on the mean and variance prediction of the Gaussian Process surrogate model.
#| fig-alt: FIXME
ggplot(aes(x = x, y = ei), data = xydt) +
  geom_line() +
  geom_point(aes(x = x, y = acq_ei), colour = "darkred", size = 2, data = candidate) +
  labs(y = "Expected Improvement") +
  theme_minimal()
```

We proceed to evaluate the candidate and would then continue with the next iteration of the loop.

```{r}
instance$eval_batch(candidate)
```

> ## Bayesian Optimization for Black Box Optimization {#sec-bayesian-black-box-optimization}

Reviewer: I think the problem with this chapter is that it takes quite a long time to get to, IMO, the most important part, which is this section. Here we finally see why we should care about everything above. I think it could benefit from having an example like this very early on to hook users

  terminator = trm("evals", n_evals = 25)
)
```

We can then simply construct an `r ref("TunerMbo")` and use it to optimize the instance.
Note that `r ref("TunerMbo")` is simply a light-weight wrapper around `r ref("OptimizerMbo")`.

```{r}
tuner = tnr("mbo",
  loop_function = bayesopt_ego,
  surrogate = surrogate,
  acq_function = acq_function,
  acq_optimizer = acq_optimizer)
```

```{r, output=FALSE}
tuner$optimize(instance)
```

```{r}
instance$result
```

We see that BO finds a substantially better hyperparameter configuration than a grid or random search (using the same budget):

```{r, output=FALSE}
instance = ti(
  task = tsk("sonar"),
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = msr("classif.ce"),
  terminator = trm("evals", n_evals = 25)
)
tuner = tnr("grid_search", resolution = 5)
tuner$optimize(instance)
```

```{r}
instance$result
```

```{r}
instance = ti(
  task = tsk("sonar"),
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = msr("classif.ce"),
  terminator = trm("evals", n_evals = 25)
)
tuner = tnr("random_search", batch_size = 25)
tuner$optimize(instance)
```

```{r, output=FALSE}
instance$result
```

> ## Multi-Objective Bayesian Optimization {#sec-multi-objective-bayesian-optimization}

Reviewer: On first read unclear what this section adds that hasn't been covered in 12.4 and C4

An example for this approach is given by the SMS-EGO algorithm [@ponweiser_2008] (implemented via the `r ref("mlr_loop_functions_smsego", "bayesopt_smsego")` loop function).

Using SMS-EGO for the multi-objective tuning problem above would look like the following:

```{r}
instance = ti(
  task = tsk("sonar"),
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = measures,
  terminator = trm("evals", n_evals = 20),
  store_models = TRUE  # required to inspect selected_features
)
```

```{r}
surrogate = srlrnc(list(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS", control = list(trace = FALSE)), lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS", control = list(trace = FALSE))))
acq_function = acqf("smsego")
acq_optimizer = acqo(opt("random_search", batch_size = 1000),
  terminator = trm("evals", n_evals = 1000))

tuner = tnr("mbo",
  loop_function = bayesopt_smsego,
  surrogate = surrogate,
  acq_function = acq_function,
  acq_optimizer = acq_optimizer)
```

> ## Noisy Bayesian Optimization {#sec-noisy-bayesian-optimization}

Reviewer: Hard to follow without an example

So far, we implicitly assumed that the black box function we are trying to optimize is deterministic, i.e., repeatedly evaluating the same point will always returns the same objective function value.
Real world black box functions, however, are often noisy, i.e., the true signal of the black box function is augmented by some noise and repeatedly evaluating the same point will return different objective function values.
In `r bbotk`, noisiness of an objective function can be indicated via the `properties` argument.
This allows users to programmatically treat such objectives differently.

```{r}
sinus_1D_noisy = function(xdt) {
  y = 2 * xdt$x * sin(14 * xdt$x) + rnorm(nrow(xdt), mean = 0, sd = 0.1)
  data.table(y = y)
}
domain = ps(x = p_dbl(lower = 0, upper = 1))
codomain = ps(y = p_dbl(tags = "minimize"))
objective = ObjectiveRFunDt$new(sinus_1D_noisy,
  domain = domain, codomain = codomain, properties = "noisy")

> objective$properties

Reviewer: ```suggestion
```

srlrn(lrn("regr.km", nugget.estim = TRUE))
```

This will result in the Gaussian Process not perfectly interpolating training data and variance being non zero for training data.
A more in-depth discussion of noise free vs. noisy observations in the context of Gaussian Processes can be found in @williams_2006.

An example for an acquisition function that properly respects noisiness of observations is given by the Augmented Expected Improvement [@huang_2012] which essentially rescales the Expected Improvement taking measurement error into account:

```{r, output=FALSE}
acqf("aei")
```

Finally, `r mlr3mbo` allows for explicitly specifying how the final result after optimization is assigned to the instance (i.e. what will be written to `instance$result`) via so-called result assigners.

```{r}
as.data.table(mlr_result_assigners)
```

For example, `r ref("mlr_result_assigners_surrogate", "ResultAssignerSurrogate")` will not simply pick the best point according to the evaluations logged in the `archive` but instead will use a surrogate model to predict the mean of all evaluated points and proceed to choose the point with the best mean prediction as the final result.

> ## Robustness, Safety Nets, Defaults and Practical Considerations

Reviewer: ```suggestion
## Practical considerations in Bayesian Optimization
```

Suboptimal parameter values can result in white noise models with a constant mean and variance prediction (except for the interpolation of training data).
In this case, the surrogate model will not provide useful mean and variance predictions resulting in poor performance of the BO algorithm.
Moreover, fitting a vanilla Gaussian Process scales cubic in the number of data points and therefore the overhead of the BO algorithm grows with the number of iterations.
Besides, vanilla Gaussian Processes natively cannot handle categorical input variables or dependencies in the search space (recall that in HPO we often deal with hierarchical mixed spaces).
In contrast, a random forest is cheap to train, quite robust in the sense that it is not that sensitive to its hyperparameters as a Gaussian Process and can easily handle hierarchical mixed spaces.
On the downside, a random forest is not really Bayesian (i.e., there is no posterior predictive distribution) and suffers from poor uncertainty estimates and poor extrapolation.

> Nevertheless, random forests usually perform quite well as a surrogate model in BO.

Reviewer: Unsure how much of this detail is really needed and useful here, it is very technical. On the other hand some readers might really like that

## Conclusion

In this chapter, we learned how to tackle black box optimization with Bayesian optimization.

> Bayesian optimization 
> `r mlr3mbo` is built modular relying on a `r ref("Surrogate")`, `r ref("AcqFunction")` and `r ref("AcqOptimizer")` as well as a general `r ref("loop_function")` that build the actual optimizer or tuner constructed in the form of an `r ref("OptimizerMbo")` or `r ref("TunerMbo")`.

Reviewer: ```suggestion
Bayesian optimization using `r mlr3mbo` relies on a `r ref("Surrogate")`, `r ref("AcqFunction")` and `r ref("AcqOptimizer")` as well as a general `r ref("loop_function")` that build the actual optimizer or tuner constructed in the form of an `r ref("OptimizerMbo")` or `r ref("TunerMbo")`.
```

> A more in-depth introduction to `r mlr3mbo` is given in its getting started vignette.

Reviewer: link to vignette
